# Exploratory data analysis
In this notebook, we will perform an extensive exploratory data analysis (EDA) in order to get more insight into the data. This will be useful for more elaborate preprocessing.

### Table of contents
* [Load data](#load)
* [First data inspection](#inspection)
* [Faulty data](#faulty)
* [Missing data](#missing)
* [Categorical variables](#categorical)
* [Distributions](#distributions)
* [Quantitative relationships & feature importance](#quantitative)
* [Qualitative relationships](#qualitative)
* [Outliers](#outliers)
* [Clusters](#clusters)

In [74]:
import pandas as pd

# display full dataframes:
pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)

train_data_file = "../data/train.csv"
target_col = "SalePrice"

## Load data <a class="anchor"  id="load"></a>
To prevent any biased decisions, the EDA is performed only on the training set. Thus, only the training data is loaded.

In [82]:
train_df = pd.read_csv(train_data_file)
train_df.set_index("Id", inplace=True)
y_train = train_df[target_col]
X_train = train_df.drop(columns=[target_col])

## First data inspection <a class="anchor"  id="inspection"></a>
**Questions: What is the size of data? What does the data look like? What are the variable data types?**

In [83]:
print(f"Size of training data: {train_df.shape}")

Size of training data: (1460, 80)


In [76]:
train_df.sample(5)

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
Id                                                                      
975           70       RL         60.0    11414   Pave   NaN      IR1   
6             50       RL         85.0    14115   Pave   NaN      IR1   
1048          20       RL         57.0     9245   Pave   NaN      IR2   
1309          20       RM        100.0    12000   Pave   NaN      Reg   
166          190       RL         62.0    10106   Pave   NaN      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
Id                                                                       
975          Lvl    AllPub    Corner       Gtl      BrkSide       RRAn   
6            Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
1048         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1309         Lvl    AllPub    Inside       Gtl      OldTown       Norm   
166          Lvl    AllPub    Inside       Gtl      Edwards       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
Id                                                                         
975       Feedr     1Fam     2Story            7            8       1910   
6          Norm     1Fam     1.5Fin            5            5       1993   
1048       Norm     1Fam     1Story            5            5       1994   
1309       Norm     1Fam     1Story            5            7       1948   
166        Norm   2fmCon     1.5Fin            5            7       1940   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
Id                                                                         
975           1993     Gable  CompShg     HdBoard     HdBoard        NaN   
6             1995     Gable  CompShg     VinylSd     VinylSd        NaN   
1048          1995     Gable  CompShg     VinylSd     VinylSd        NaN   
1309          2005     Gable  CompShg     MetalSd     MetalSd        NaN   
166           1999     Gable  CompShg     Wd Sdng     Wd Sdng        NaN   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
Id                                                                  
975          0.0        TA        Gd     BrkTil       Gd       TA   
6            0.0        TA        TA       Wood       Gd       TA   
1048         0.0        TA        TA      PConc       Gd       TA   
1309         0.0        TA        TA     CBlock       TA       TA   
166          0.0        TA        Gd     BrkTil       TA       TA   

     BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  \
Id                                                                    
975            No          Unf           0          Unf           0   
6              No          GLQ         732          Unf           0   
1048           No          GLQ         686          Unf           0   
1309           No          GLQ         144          ALQ         608   
166            No          ALQ         351          Rec         181   

      BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  \
Id                                                                     
975         728          728    GasA        TA          N      SBrkr   
6            64          796    GasA        Ex          Y      SBrkr   
1048        304          990    GasA        Ex          Y      SBrkr   
1309        172          924    GasA        Ex          Y      SBrkr   
166         112          644    GasA        Gd          Y      SBrkr   

      1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
Id                                                                              
975       1136       883             0       2019             0             0   
6          796       566             0       1362             1             0   
1048       990         0             0        990             0             1   
1309      1122   

In [77]:
# distribution of data types:
X_train.dtypes.value_counts()

object     43
int64      33
float64     3
Name: count, dtype: int64

In [78]:
print(f"Data type of target: {y_train.dtype}")

Data type of target: int64


**Findings**:
The training data has 1460 rows and 80 columns. One of these columns is the target. Consequently, there are 79 feature columns where 43 are categorical, 33 integers and 3 floating point numbers. The target is represented as integer.

## Missing data <a class="anchor"  id="missing"></a>
**Questions: How is missing data represented? How much data is missing? Do missing values actually mean a specific value?**

## Faulty data <a class="anchor"  id="faulty"></a>
**Questions: Are there duplicated rows or columns? Does the feature’s data type make sense? Is a numerical feature actually categorical or vice versa? Do all values of one feature have the same data type? Are all values in the expected range? If multiple features are connected, is there any inconsistency?**

## Categorical variables <a class="anchor"  id="categorical"></a>
**Questions: Is there an ordering in the categories? Are there special characters (non ASCII)?**

## Distributions <a class="anchor"  id="distributions"></a>
**Questions: Is the feature continuous or discrete? If discrete: How many distinct values are there? Is the distribution similar to a normal distribution? Is the distribution skewed? Are there particularly frequent or rare values? Are there inf values? Do the features have a similar range? Are the assumptions of the ML algorithm you want to use met?**

## Quantitative relationships & feature importance <a class="anchor"  id="quantitative"></a>
**Questions: Are there collinear variables? Which features are most predictive?**

## Qualitative relationships <a class="anchor"  id="qualitative"></a>
**Questions: Is there a clear relationship of a feature with the target that has not been caught by the quantitative analysis? Is a transformation of a feature helpful with your ML algorithm?**

## Outliers <a class="anchor"  id="outliers"></a>
**Questions: Are there outliers? Are those outliers misleading the prediction?**

## Clusters <a class="anchor"  id="clusters"></a>
**Questions: Are there clear clusters in the data that should be predicted separately?**